pkg/cmd/cli/cmd/cancelbuild.go

```diff
@@ -3,166 +3,238 @@ package cmd
-	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/util/wait"
 
 	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	osclient "github.com/openshift/origin/pkg/client"
+	cmdutil "github.com/openshift/origin/pkg/cmd/util"
 	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
 )
 
 const (
 	cancelBuildLong = `
-Cancels a pending or running build
+Cancel a running, pending, or new builds
 
-This command requests a graceful shutdown of the running build. There may be a delay between requesting 
+This command requests a graceful shutdown of the build. There may be a delay between requesting 
 the build and the time the build is terminated.`
 
 	cancelBuildExample = `  # Cancel the build with the given name
-  $ %[1]s cancel-build 1da32cvq
+  $ %[1]s cancel-build ruby-build-2
 
   # Cancel the named build and print the build logs
-  $ %[1]s cancel-build 1da32cvq --dump-logs
+  $ %[1]s cancel-build ruby-build-2 --dump-logs
 
   # Cancel the named build and create a new one with the same parameters
-  $ %[1]s cancel-build 1da32cvq --restart`
+  $ %[1]s cancel-build ruby-build-2 --restart
+
+  # Cancel multiple builds
+  $ %[1]s cancel-build ruby-build-1 ruby-build-2 ruby-build-3
+
+  # Cancel all builds created from 'ruby-build' build configuration that are in 'new' state
+  $ %[1]s cancel-build bc/ruby-build --state=new`
 )
 
+type CancelBuildOptions struct {
+	In          io.Reader
+	Out, ErrOut io.Writer
+
+	DumpLogs   bool
+	Restart    bool
+	State      string
+	Namespace  string
+	BuildNames []string
+
+	Mapper      meta.RESTMapper
+	Client      osclient.Interface
+	BuildClient osclient.BuildInterface
+}
+
 // NewCmdCancelBuild implements the OpenShift cli cancel-build command
-func NewCmdCancelBuild(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+func NewCmdCancelBuild(fullName string, f *clientcmd.Factory, in io.Reader, out io.Writer) *cobra.Command {
+	o := &CancelBuildOptions{}
+
 	cmd := &cobra.Command{
-		Use:        "cancel-build BUILD",
-		Short:      "Cancel a pending or running build",
+		Use:        "cancel-build (BUILD | BUILDCONFIG)",
+		Short:      "Cancel a running, pending, or new builds",
 		Long:       cancelBuildLong,
 		Example:    fmt.Sprintf(cancelBuildExample, fullName),
 		SuggestFor: []string{"builds", "stop-build"},
 		Run: func(cmd *cobra.Command, args []string) {
-			err := RunCancelBuild(f, out, cmd, args)
-			kcmdutil.CheckErr(err)
+			kcmdutil.CheckErr(o.Complete(f, in, out, cmd, args))
+			kcmdutil.CheckErr(o.Run())
 		},
 	}
 
-	cmd.Flags().Bool("dump-logs", false, "Specify if the build logs for the cancelled build should be shown.")
-	cmd.Flags().Bool("restart", false, "Specify if a new build should be created after the current build is cancelled.")
-	//cmdutil.AddOutputFlagsForMutation(cmd)
+	cmd.Flags().StringVar(&o.State, "state", o.State, "Specify the state of the builds to be cancelled")
```

Comment:
i'd change this description to "Only cancel builds in this state".

otherwise a confused person might read the description as "builds that are canceled by this command will end up in the state specified by --state"